graph/src/main/scala/net/spals/dsa/graph/Graph.scala

```diff
@@ -0,0 +1,57 @@
+package net.spals.dsa.graph
+
+/**
+  * Functional implementation of a weighted graph.
+  *
+  * @author spags
+  * @author tkral
+  */
+case class Edge[T, W](v1: Vertex[T], v2: Vertex[T], w: W) {  }
+
+case class Vertex[T](v: T) {  }
+
+case class WeightedGraph[T, W](vertices: Set[Vertex[T]], edgeFunc: (Vertex[T], Vertex[T]) => Option[W]) {
+
+  def addEdge(v1: Vertex[T], v2: Vertex[T], w: W): WeightedGraph[T, W] = {
+    val addEdgeFunc: (Vertex[T], Vertex[T]) => Option[W] =
+      (vx, vy) => (vx, vy) match {
+        case (vx, vy) if (vx == v1 && vy == v2) || (vx == v2 && vy == v1) => Option(w)
+        case _ => edgeFunc.apply(vx, vy)
+      }
+    WeightedGraph(vertices ++ Set(v1, v2), addEdgeFunc)
```

Comment:
Is this the correct semantics for a graph? Or should we reject if the vertices don't exist in the graph?